### PR TITLE
Looks like the attention mechanism works. Would need a review, though…

### PR DESCRIPTION
**Pull request by postnubilaphoebus**
…. Inspect especially lines 140-160 in models.py. I just uploaded the notebook to see the results. It is old can be discarded during merge.

### CODE REVIEW

File: src/models.py [hunk 1]
```diff
@@ -16,24 +24,69 @@
...
+        self.pretrained.layer4[2].bn2.register_forward_hook(self.forward_hook('BatchNorm'))
+        
+    def forward_hook(self,layer_name):
+        def hook(module, input, output):
+            self.intermediate.update({layer_name: output})
+        return hook
+
+    def forward(self, input):
+        out = self.pretrained(input)
+        return out, self.intermediate
```
Comment (tobiasvanderwerff) [inline src/models.py]: No need to return `self.intermediate`, since this is already an accesible property of the ResNet class.

```suggestion
        return out
```

File: src/models.py [hunk 2]
```diff
@@ -16,24 +24,69 @@
...
+        self.mlp2 = MLP(decoder.num_hidden)
         self.decoder = decoder
         
     def forward(self, img, caption, seq_lengths, targets=None):
-        img_features = self.encoder(img)  # shape: (batch, img_features)
-        img_features = img_features.repeat(2, 1, 1)  # shape: (num_layers, batch, img_features)
-#         import pdb; pdb.set_trace()
-        logits = self.decoder(caption, img_features, seq_lengths) 
-        
+        img_features, intermediate_layer1 = self.encoder(img)
```
Comment (tobiasvanderwerff) [inline src/models.py]: This connects to my comment of line 56 -- you could access the intermediate layer immediately:
```suggestion
        img_features = self.encoder(img)
        intermediate_layer1 = self.encoder.intermediate
```

``

File: src/models.py [hunk 3]
```diff
@@ -16,24 +24,69 @@
...
         self.decoder = decoder
         
     def forward(self, img, caption, seq_lengths, targets=None):
-        img_features = self.encoder(img)  # shape: (batch, img_features)
-        img_features = img_features.repeat(2, 1, 1)  # shape: (num_layers, batch, img_features)
-#         import pdb; pdb.set_trace()
-        logits = self.decoder(caption, img_features, seq_lengths) 
-        
+        img_features, intermediate_layer1 = self.encoder(img)  
+        #intermediate layer shape: [2, 512, 4, 4]
```
Comment (tobiasvanderwerff) [inline src/models.py:70]: Don't hardcode the shape of the intermediate layer, since some of these shape parameters can change depending on the model parameters. For example, you could make it `[batch, 512, 4, 4]` (not sure if this is the correct size, but hopefully the idea is clear). I believe the last three shape parameters are a property of the predefined architecture of the ResNet, so those should remain static.

File: src/models.py [hunk 4]
```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
         self.device = device
         
         self.emb = nn.Embedding(vocab_size, embedding_dim)
-        self.lstm = nn.LSTM(embedding_dim, num_hidden, num_layers=2, batch_first=True, dropout=0, bidirectional=bidirectional)
-        self.fc = nn.Linear(num_hidden, vocab_size)
+        self.lstm = nn.LSTM(embedding_dim, num_hidden, num_layers=2, batch_first=True, dropout=0, bidirectional=bidirectional) 
+        self.attn = nn.Linear(self.num_hidden * 2, 1)
```
Comment (tobiasvanderwerff) [inline src/models.py:103]: It's unfortunate that in the Show, Attend and Tell paper they say that they use a multilayer perceptron for the attention function, but they don't really specify the details, e.g. the amount of hidden layers. My intuition is that we might want to turn this into two linear layers chained together. The reason why I think this is that the attention mechanism should model some sort of relationship between the two stacked input vectors (the decoder hidden state and one of the annotation vectors). However, in a single linear layer you cannot consider the relationship between different values of the input, which _would_ be possible if you add another linear layer after it. 

The same thing goes for the MLP class at the beginning of this file. It's unclear how many layers it should have. My proposal would be to also add a second layer there, and see how well it works.

File: src/models.py [hunk 5]
```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
...
+        self.fc = nn.Linear(self.num_hidden * 2, vocab_size)
     
-    def forward(self, x, h0, seq_lengths):
-        # x has shape (batch, seq_len), h0 has shape (num_layers * num_directions, batch, num_hidden)
+    def forward(self, caption, h0, c0, seq_lengths, intermediate_layer):
+        # -------------- Inputs --------------------
+            # caption = caption; shape = (batch, seq_len)
+            # mlp = image feature vector averaged and passed through MLP
+            # seq_lengths = ''
+            # intermediate_layer = output of 'encoder.layer4[2].bn2'
```
Comment (tobiasvanderwerff) [inline src/models.py]: Small detail, but you should use triple quotes (""") for the docstring at the beginning of a function instead of the # comments, i.e. 

```suggestion
"""  
          -------------- Inputs --------------------  
             caption = caption; shape = (batch, seq_len)   
             mlp = image feature vector averaged and passed through MLP   
             seq_lengths = ''   
             intermediate_layer = output of 'encoder.layer4[2].bn2'    
"""
```

File: src/models.py [hunk 6]
```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
...
+        emb = self.emb(caption)  # (batch, seq_len) -> (batch, seq_len, embedding_dim)
+     
         # Pack_padded_sequence so that padded items in the sequence won't be shown to the LSTM
         # See https://towardsdatascience.com/taming-lstms-variable-sized-mini-batches-and-why-pytorch-is-good-for-your-health-61d35642972e
         # We use seq_lengths - 1 because we do not want to count the last <END> token in each caption.
         emb = torch.nn.utils.rnn.pack_padded_sequence(emb, seq_lengths - 1, batch_first=True)
+
+        # [2, 38, 512]
+
+        # input of shape (seq_len, batch, input_size):
```
Comment (tobiasvanderwerff) [inline src/models.py]: Should lines 123-126 be removed? Not sure what they are referring to

```suggestion
```

File: src/models.py [hunk 7]
```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
...
         
         out, _ = self.lstm(emb, (h0, c0))
 
         # Undo the packing operation
         out, _ = torch.nn.utils.rnn.pad_packed_sequence(out, batch_first=True)  # (batch, seq_len, num_directions * num_hidden)
         
-        logits = self.fc(out)
+        # ---------------- Attention -----------------------
+
+        # torch.Size([2, 512, 16]) -> torch.Size([2, 16, 512])
```
Comment (tobiasvanderwerff) [inline src/models.py]: Same thing as for line 69: hardcoded shapes

```suggestion
        # torch.Size([batch, num_hidden, n_annotations]) -> torch.Size([batch, n_annotations, num_hidden])
```

File: src/models.py [hunk 8]
```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
...
+        intermediate_layer = torch.transpose(intermediate_layer, 1, -1)
+        
+        attn_in = torch.zeros(out.size(0), out.size(1), out.size(2) * 2).to(self.device)
```
Comment (tobiasvanderwerff) [inline src/models.py:139]: What it looks like to me is that for `attn_in`, you want to take the shape of `out`, and add another dimension to it that can fit all the 16 combinations of hidden state + one of the 16 annotation vectors, i.e. `attn_in.shape = [batch, seq_len, 16, num_hidden * 2]`. Then given a batch number `b`, timestep `t`, along with the `k'th` annotation vector `ann` and the hidden state `h` at batch `b` and time `t`, you have `attn_in[b, t, k, :] =  torch.cat(ann, h)`. In code:

```
# intermediate_layer has shape [batch, n_annotations, num_hidden],
# out has shape [batch, seq_len, num_hidden].

n_annotations = intermediate_layer.size(1)
attn_in = torch.zeros(batch, seq_len, n_annotations, num_hidden * 2)
for b in range(batch):
   for t in range(seq_len):
      h = out[b, t]
      for k, ann in enumerate(intermediate_layer[b]):  # iterate over all annotation vectors, 16 in this example
         attn_in[b, t, k, :] = torch.cat(ann, h)
```

You should start with writing this out in a for loop as above, because it's easier to verify what you're doing this way. You should only consider making it more efficient by replacing (part of) the for loops with commands like `torch.cat` once you are sure that you understand the loopy version properly.

File: src/models.py [hunk 9]
```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
...
         emb = torch.nn.utils.rnn.pack_padded_sequence(emb, seq_lengths - 1, batch_first=True)
+
+        # [2, 38, 512]
+
+        # input of shape (seq_len, batch, input_size):
         
         out, _ = self.lstm(emb, (h0, c0))
 
         # Undo the packing operation
         out, _ = torch.nn.utils.rnn.pad_packed_sequence(out, batch_first=True)  # (batch, seq_len, num_directions * num_hidden)
```
Comment (tobiasvanderwerff) [inline src/models.py]: Since you reuse the shape parameters of the `out` and `intermediate_layer` tensor several times, I would assign them to variables. This also helps readability, since it's then easier to see what the shapes represent. For example,  you can add 

```suggestion

batch, seq_len, directions_times_hidden = out.shape
_, _, n_annotations = intermediate_layer.shape
```

Then  you can also use these variables from now on, because for me it can be hard to read when you constantly  refer back to the shape parameters of for example `out` or `intermediate_layer`, which I then have to look up.

File: src/models.py [hunk 10]
```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
...
+        
+        for i in range(out.size(0)):
+            seq_len = seq_lengths[i].item() - 1
+
+            stack_intermediate_layer = torch.zeros(out.size(1), out.size(2)).to(self.device)
+
+            # As the two tensors (out, intermediate_layer) have different shapes,
+            # I copy from intermediate_layer until I have reached shape [2, 38, 512] ('out' shape)
+            for j in range(seq_len):
+                idx = j % (intermediate_layer.size(1) - 1)
```
Comment (tobiasvanderwerff) [inline src/models.py:149]: This somehow feels wrong to me. Why would it be necessary to take the modulo of `j`? If `seq_len > (intermediate_layer.size(1) - 1)`, i.e. in this case `seq_len > 14`, this would mean that `j = 1,2,...,14,1,2...`. Since `j` repeats itself, wouldn't you be overwriting previous results in line 149? See also my comment on line 138.

File: src/models.py [hunk 11]
```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
...
+                stack_intermediate_layer[j] = intermediate_layer[i][idx].clone().detach()
+
+            attn_in[i, :seq_len] = torch.cat((stack_intermediate_layer, out[i, :seq_len]), dim = 1)
+
+        attn_out = self.attn(attn_in)
+        attn_weights = F.softmax(attn_out, dim=1)
```
Comment (tobiasvanderwerff) [inline src/models.py]: Make sure you use the right dimension for the softmax -- you want to do it over the `n_annotations` dimension. If `attn_out` has shape `[batch, seq_len, n_annotations, 1]`, you should use `attn_out.squeeze()` to remove the redundant last dimension, and then call softmax on the last dimension: `F.softmax(attn_out, dim=-1)`.

```suggestion
        attn_out = self.attn(attn_in).squeeze()
        attn_weights = F.softmax(attn_out, dim=-1)
```

File: src/models.py [hunk 12]
```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
...
+        
+        cs = torch.sum(attn_weights.unsqueeze(1) * out, dim=1, keepdim=False)
```
Comment (tobiasvanderwerff) [inline src/models.py:157]: Once again, pay attention to the dimension. Are you not summing over the LSTM timesteps?

**Review by tobiasvanderwerff — COMMENTED** [review]
Nice job. I would say the main thing that is not yet quite right is the part where you prepare the input for the attention mechanism; this still seems to require some changes.

File: src/models.py (hunk 2 above)
Comment (postnubilaphoebus) [reply src/models.py]: ok

File: src/models.py (hunk 3 above)
Comment (postnubilaphoebus) [reply src/models.py:70]: Line 69 was just a comment, so I'm not sure what you are referring to. I can change the comment to [batch, 512, 4, 4] or even [batch, num_hidden, dim1_featmap, dim2_featmap].

File: src/models.py (hunk 4 above)
Comment (postnubilaphoebus) [inline src/models.py:103]:
```suggestion
        self.attn = nn.Linear(self.num_hidden * 2, 1) 
        self.attn2 = nn.Linear(self.num_hidden * 2, 1) 
```

Comment (postnubilaphoebus) [reply src/models.py:103]: Good point, I proposed that change!

File: src/models.py (hunk 6 above)
Comment (postnubilaphoebus) [reply src/models.py]: yes, remained from debugging

File: src/models.py (hunk 11 above)
Comment (postnubilaphoebus) [reply src/models.py]: Thanks for spotting that. I was not aware of the dimension change.

File: src/models.py [hunk 13]
```diff
@@ -16,24 +24,69 @@
 
 logger = logging.getLogger(__name__)
 
+class MLP(nn.Module):
+    def __init__(self, num_hidden):
+        super().__init__()
+        self.num_hidden = num_hidden
+
+        self.lin1 = nn.Linear(num_hidden, num_hidden)
```
Comment (postnubilaphoebus) [inline src/models.py:32]:
```suggestion
        self.lin1 = nn.Linear(num_hidden, num_hidden)
        self.relu1 = nn.ReLU(inplace=True)
        self.lin2 = nn.Linear(num_hidden, num_hidden)
        self.relu2 = nn.ReLU(inplace=True)
```
Replace one layer MLP with two layer MLP.

File: src/models.py [hunk 14]
```diff
@@ -16,24 +24,69 @@
...
+        self.relu1 = nn.ReLU(inplace=True)
```
Comment (postnubilaphoebus) [inline src/models.py:33]:
```suggestion
```

File: src/models.py [hunk 15]
```diff
@@ -16,24 +24,69 @@
...
+       
+    def forward(self, x):
+        x = self.lin1(x)
+        x = self.relu1(x)
```
Comment (postnubilaphoebus) [inline src/models.py:37]:
```suggestion
        x = self.relu1(x)
        x = self.lin2(x)
        x = self.relu2(x)
```

Comment (postnubilaphoebus) [reply src/models.py:37]: Replace one layer MLP with two layer MLP.

File: src/models.py (hunk 3 above)
Comment (postnubilaphoebus) [inline src/models.py:70]:
```suggestion
        #intermediate layer shape: [batch, num_hidden, dim1_featmap, dim2_featmap]
```

File: src/models.py (hunk 10 above)
Comment (postnubilaphoebus) [reply src/models.py:149]: I'm not overwriting previous results because 'idx' is the result of the modulo operation, not 'j'.
I was just gonna store multiple versions of the same information because I was unsure about how to tackle the dimension problem. Your suggestion in line 138 made things clearer, though.

File: src/models.py (hunk 8 above)
Comment (postnubilaphoebus) [reply src/models.py:139]: I see, I didn't think about adding another dimension. That makes it easier.

Comment (postnubilaphoebus) [inline src/models.py:139]:
```suggestion
# intermediate_layer has shape [batch, n_annotations, num_hidden],
# out has shape [batch, seq_len, num_hidden].

n_annotations = intermediate_layer.size(1)
attn_in = torch.zeros(batch, seq_len, n_annotations, num_hidden * 2).to(self.device)
for b in range(batch):
   for t in range(seq_len):
      h = out[b, t]
      for k, ann in enumerate(intermediate_layer[b]):  # iterate over all annotation vectors, 16 in this example
         attn_in[b, t, k, :] = torch.cat(ann, h)
```
I would just take your suggestion. It's nicely readable and tbh I don't want to shorten it to the point where it would be difficult to see what's going on.

File: src/models.py (hunk 10 above)
Comment (postnubilaphoebus) [inline src/models.py:149]:
```suggestion
       
```
Delete in favor of above code

File: src/models.py [hunk 16]
```diff
@@ -45,38 +98,78 @@ def __init__(self, num_hidden, embedding_dim, vocab_size, device, num_layers=2,
...
+
+            stack_intermediate_layer = torch.zeros(out.size(1), out.size(2)).to(self.device)
+
+            # As the two tensors (out, intermediate_layer) have different shapes,
+            # I copy from intermediate_layer until I have reached shape [2, 38, 512] ('out' shape)
+            for j in range(seq_len):
+                idx = j % (intermediate_layer.size(1) - 1)
+                stack_intermediate_layer[j] = intermediate_layer[i][idx].clone().detach()
+
+            attn_in[i, :seq_len] = torch.cat((stack_intermediate_layer, out[i, :seq_len]), dim = 1)
```
Comment (postnubilaphoebus) [inline src/models.py:152]:
```suggestion
               
```
delete in favor of above code (proposed change between 138 and 149)

File: src/models.py (hunk 12 above)
Comment (postnubilaphoebus) [reply src/models.py:157]: Should it be

```suggestion
        cs = torch.sum(attn_weights.unsqueeze(1) * out, dim=2, keepdim=False)
```
because we are summing over the hidden layers? I tried that, but got a bug for some reason.

File: src/models.py (hunk 11 above)
Comment (postnubilaphoebus) [inline src/models.py]:
```suggestion
        attn_out = self.attn(attn_in).squeeze()
        attn_weights = F.softmax(attn_out, dim=-1)
        attn_out = self.attn2(attn_weights).squeeze()
        attn_weights = F.softmax(attn_out, dim=-1)
```

Comment (postnubilaphoebus) [reply src/models.py]: because we have two attention layers now?

**Review by postnubilaphoebus — COMMENTED** [review]
Thanks for the review! I hope this improves things.

File: src/models.py (hunk 12 above)
Comment (tobiasvanderwerff) [reply src/models.py:157]: Yes, I indeed meant it should be dim=2
